src/transpose_z_to_y.f90

```diff
@@ -476,4 +429,182 @@ subroutine mem_merge_zy_complex(in, n1, n2, n3, out, iproc, dist, decomp)
+#if defined(_GPU)
+      attributes(device) :: wk1, wk2
+#endif
+     
+#ifdef EVEN
+      call define_recvbuf_real_even(src, decomp, wk1, wk2)
+#else
+
+#if defined(_GPU)
+      call define_recvbuf_real_gpu(decomp, wk1, wk2)
+#else
+      associate (wk => wk1)
+      end associate
+      call define_recvbuf_real_std(src, decomp, wk2)
+#endif
+#endif
+
+   end subroutine define_recvbuf_real
+
+   subroutine rearrange_recvbuf_real(src, dst, decomp, wk2)
+
+      implicit none
+     
+      real(mytype), dimension(:, :, :), intent(IN) :: src
+      real(mytype), dimension(:, :, :), intent(OUT) :: dst
+      TYPE(DECOMP_INFO), intent(IN) :: decomp
+      real(mytype), dimension(:), intent(out) :: wk2
+#if defined(_GPU)
+      attributes(device) :: wk2
+#endif
+
+      integer :: s1, s2, s3, d1, d2, d3
+
+      s1 = SIZE(src, 1)
+      s2 = SIZE(src, 2)
+      s3 = SIZE(src, 3)
+      d1 = SIZE(dst, 1)
+      d2 = SIZE(dst, 2)
+      d3 = SIZE(dst, 3)
+
+      call mem_merge_zy_real(wk2, d1, d2, d3, dst, dims(2), &
+           decomp%y2dist, decomp)
+      
+   end subroutine rearrange_recvbuf_real
+
+!!!! Option-specific utility subroutines
+   
+#ifndef EVEN
+   
+#ifndef _GPU
+
+   subroutine define_recvbuf_real_std(src, decomp, wk2)
+
+      implicit none
+
+      real(mytype), dimension(:, :, :), intent(IN) :: src
+      TYPE(DECOMP_INFO), intent(IN) :: decomp
+      real(mytype), dimension(:), intent(out) :: wk2
+
+      integer :: ierror
+
+      call MPI_ALLTOALLV(src, decomp%z2cnts, decomp%z2disp, real_type, &
+                         wk2, decomp%y2cnts, decomp%y2disp, real_type, &
+                         DECOMP_2D_COMM_ROW, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLTOALLV")
+      
+   end subroutine define_recvbuf_real_std
+
+#else
+   ! _GPU is defined
+   subroutine define_recvbuf_real_gpu(decomp, wk1, wk2)
+
+      implicit none
+
+      TYPE(DECOMP_INFO), intent(IN) :: decomp
+      real(mytype), dimension(:), intent(out) :: wk1, wk2
+      attributes(device) :: wk1, wk2
+
+#ifndef _NCCL
+      integer :: istat
+      integer :: ierror
+      
+      call MPI_ALLTOALLV(wk1, decomp%z2cnts, decomp%z2disp, real_type, &
+                         wk2, decomp%y2cnts, decomp%y2disp, real_type, &
+                         DECOMP_2D_COMM_ROW, ierror)
+      if (ierror /= 0) call decomp_2d_abort(__FILE__, __LINE__, ierror, "MPI_ALLTOALLV")
+#else
+      call decomp_2d_nccl_send_recv_row(wk2, &
+                                        wk1, &
+                                        decomp%z2disp, &
+                                        decomp%z2cnts, &
+                                        decomp%y2disp, &
+                                        decomp%y2cnts, &
+                                        dims(2))
+#endif
+
+   end subroutine define_recvbuf_real_gpu
+
+   ! END IFDEF _GPU
+#endif
+
+#else
+   ! EVEN is defined
+
+   subroutine define_recvbuf_real_even(src, decomp, wk1, wk2)
```

Comment:
This subroutine could be always defined ? Is it important to surround it with preprocessor macros ?

Reply:
Yes - it is unused if `EVEN` is not defined and the compiler will complain